File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Main {

    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static int N, S, ANS = 0;
    private static Integer[] ARR;

    private static List<Integer> visited = new ArrayList<>();

    public static void main(String[] args) throws IOException {
        String[] input = br.readLine().split(" ");
        N = Integer.parseInt(input[0]);
        S = Integer.parseInt(input[1]);
        ARR = Arrays.stream(br.readLine().split(" "))
                .map(Integer::parseInt)
                .toArray(Integer[]::new);

        // ARR 를 백트래킹으로 탐색해보자.
        int idx = 0;
        dfs(idx, 0);

        System.out.println(ANS);
    }

    private static void dfs(int idx, int sum) {
        if (visited.size() == N) {
            boolean allZero = visited.stream()
                    .allMatch(v -> v == 0);
            if (allZero) return;

            if (sum == S) {
                ANS += 1;
            }
            return;
        }

        // idx 가 부분수열에 포함
        visited.add(1);
        dfs(idx + 1, sum + ARR[idx]);
        visited.remove(visited.size() - 1);

        // idx 가 부분수열에 미포함
        visited.add(0);
        dfs(idx + 1, sum);
        visited.remove(visited.size() - 1);
    }
}

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.py
import sys
input = sys.stdin.readline

n, s = map(int, input().split())
arr = list(map(int, input().split()))
visited = []
count = 0

def dfs(idx, sum):
    global count

    # 마지막 값을 넘어갔을 경우(즉, 이미 visited 가 n개가 전부 찼을 경우)
    if idx == n:
        # 크기가 양수인 부분수열 이므로, 전체가 False 이면 return
        # 이 부분은 매번 visited 를 전체 순회하지 않도록 최적화 가능함.(단 N <= 20 이므로 빠르게 구현)
        if not any(visited):
            return;

        if sum == s:
            count += 1
        return

    # idx 를 포함 하는 경우
    visited.append(True)
    dfs(idx + 1, sum + arr[idx])
    visited.pop()

    # idx 를 포함하지 않는 경우
    visited.append(False)
    dfs(idx + 1, sum)
    visited.pop()

def main():
    dfs(0, 0)
    print(count)

if __name__ == "__main__":
    main()